gobblin-data-management/src/main/java/gobblin/data/management/retention/action/RetentionAction.java

```diff
@@ -0,0 +1,77 @@
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.data.management.retention.action;
+
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.hadoop.fs.FileSystem;
+
+import com.typesafe.config.Config;
+
+import gobblin.data.management.dataset.Dataset;
+import gobblin.data.management.policy.VersionSelectionPolicy;
+import gobblin.data.management.version.DatasetVersion;
+import gobblin.data.management.version.finder.VersionFinder;
+import gobblin.dataset.DatasetsFinder;
+
+
+/**
+ * An abstraction to perform a retention action for a subset of {@link DatasetVersion}s.
+ * A few kinds of actions are deletion, access control, encryption, archival etc.
+ */
+public abstract class RetentionAction {
+
+  protected final FileSystem fs;
+
+  public RetentionAction(Config actionConfig, FileSystem fs) {
+    this.fs = fs;
+  }
+
+  /**
+   * Execute the action on all {@link DatasetVersion}s or a subset of {@link DatasetVersion}s. Each {@link Dataset}
+   * uses the {@link VersionFinder} to find all the {@link DatasetVersion}s and calls this method to perform the necessary
+   * action on those {@link DatasetVersion}s
+   * <p>
+   * <b>Note</b> Any kind of {@link VersionSelectionPolicy} has <b>NOT</b> been applied to the list of {@link DatasetVersion}s
+   * being passed. It is the responsibility of the {@link RetentionAction} to filter the {@link DatasetVersion}s by
+   * applying {@link VersionSelectionPolicy}s and then perform the action.
+   * </p>
+   * @param allVersions list of all {@link DatasetVersion}s found by the {@link DatasetsFinder}.
+   */
+  public abstract void execute(List<DatasetVersion> allVersions) throws IOException;
+
+  /**
+   * A factory to create new {@link RetentionAction}s
+   *
+   */
+  public interface RetentionActionFactory {
+
+    /**
+     * A factory method to create a new {@link RetentionAction} using a <code>config</code>. The {@link Dataset} always
+     * calls {@link #canCreateWithConfig(Config)} before calling this method.
+     *
+     * @param config to use to create the {@link RetentionAction}
+     * @return A new {@link RetentionAction}
+     */
+    RetentionAction createRetentionAction(Config config, FileSystem fs);
+
+    /**
+     * Method to check if a {@link RetentionAction} can be created/instantiated with the <code>config</code>.
+     * If the method returns <code>true</code>, {@link #createRetentionAction(Config, FileSystem)} can be called
+     *
+     * @param config to use to create the {@link RetentionAction}
+     * @return true if it is safe to call {@link #createRetentionAction(Config, FileSystem)}, false otherwise
+     */
+    boolean canCreateWithConfig(Config config);
```

Comment:
In general, I'd prefer the canCreateWithConfig() method because we should not be using exception for flow control. In this specific case, I don't understand why we need this method. Are there other ways to instantiate a retention action? The code below seems to ignore the "false" case which is bad.